8369432: Add Support for JDBC 4.5 MR

LanceAndersen opened the pull request:
This PR adds support for the upcoming JDBC 4.5 MR which provides the following  updates to the JDBC specification:

- Deprecate SQLPermission for removal
- Enhance the Blob/Clob/Array/SQLXML/NClob interfaces to extend/support AutoClosable
- Add the SQL types DECFLOAT, JSON to Types.Java and JDBCType.java
- Add the quoted identifier methods that **were added previously to the Statement interface in JDK 9** to the Connection interface
  - It is the exact same verbiage & default methods used when these methods were added to the Statement interface
- Clarify the Timestamp::hashCode method which incorrectly indicates that nanos are not used when calculating the hash

Tiers 1-3 have been run

LanceAndersen commented:
/csr needed

fandreuz commented on src/java.sql/share/classes/java/sql/Array.java:
```diff
@@ -61,13 +61,19 @@
  */
 
-public interface Array {
+public interface Array  extends AutoCloseable {
```
```suggestion
public interface Array extends AutoCloseable {
```

RogerRiggs commented on src/java.sql/share/classes/java/sql/Connection.java:
```diff
@@ -1676,5 +1677,277 @@ default void setShardingKey(ShardingKey shardingKey, ShardingKey superShardingKe
     default void setShardingKey(ShardingKey shardingKey) throws SQLException {
         throw new SQLFeatureNotSupportedException("setShardingKey not implemented");
     }
+    // JDBC 4.5
```
Would this be information to developers if added in and \@apiNote as a function in supported by JDBC 4.5.

turbanoff commented on src/java.sql/share/classes/java/sql/Connection.java:
```diff
@@ -1676,5 +1677,277 @@ default void setShardingKey(ShardingKey shardingKey, ShardingKey superShardingKe
+            throw new SQLException("Invalid name");
+        }
+        if (Pattern.compile("[\\p{Alpha}][\\p{Alnum}_]*").matcher(identifier).matches()) {
+            return alwaysQuote ?  "\"" + identifier + "\"" : identifier;
```
```suggestion
            return alwaysQuote ? "\"" + identifier + "\"" : identifier;
```

turbanoff commented on src/java.sql/share/classes/java/sql/Connection.java:
```diff
@@ -1676,5 +1677,277 @@ default void setShardingKey(ShardingKey shardingKey, ShardingKey superShardingKe
+     * @since 26
+     */
+    default String enquoteNCharLiteral(String val)  throws SQLException {
+        return "N'" + val.replace("'", "''") +  "'";
```
```suggestion
        return "N'" + val.replace("'", "''") + "'";
```

turbanoff commented on test/jdk/java/sql/testng/test/sql/ConnectionTests.java:
```diff
@@ -0,0 +1,228 @@
+            {"Batmobile1966", true},
+            {maxIdentifier, true},
+            {maxIdentifier + 'a', false},
+            {"", false},};
```
```suggestion
            {"", false},
        };
```

eirbjo commented:
> Pattern::matches invokes Pattern::compile and String::matches invokes Pattern::matches so did not see the benefit when we added the methods originally in JDK 9

Yeah, it just seemed `String:matches` would be less verbose. I then noticed Roger raising related suggestions about precomplication, which is why I removed my comment from the PR since it seemed no longer relevant.

LanceAndersen commented:
> > Pattern::matches invokes Pattern::compile and String::matches invokes Pattern::matches so did not see the benefit when we added the methods originally in JDK 9
> 
> Yeah, it just seemed `String:matches` would be less verbose. I then noticed Roger raising related suggestions about precomplication, which is why I removed my comment from the PR since it seemed no longer relevant.

Ah that explains why I did not see your comment...

LanceAndersen commented on src/java.sql/share/classes/java/sql/Statement.java:
```diff
@@ -1643,7 +1623,7 @@ default boolean isSimpleIdentifier(String identifier) throws SQLException {
-    default String enquoteNCharLiteral(String val)  throws SQLException {
-        return "N'" + val.replace("'", "''") +  "'";
+    default String enquoteNCharLiteral(String val) throws SQLException {
+        return SQLUtils.enquoteNCharLiteral(val);
```
Hi,

Thank you for your suggestion, but I have no plans to make the suggested change:

- The methods were added to the JDBC as of JDK 9, ~10 years ago
 - drivers that have implemented  these methods would have already added them to their Statement implementation
 - Most implementations of Statement would have more efficient access to the Connection object should they decide to move their implementation to their Connection implementation 
 -  It would be trivial to refactor these methods to their Connection implementation or another package-private class

AlanBateman commented on src/java.sql/share/classes/java/sql/package-info.java:
```diff
@@ -166,6 +173,26 @@
  *      </UL>
  * </UL>
  *
+ *     <h3>{@code java.sql} and {@code javax.sql} Features Introduced in the JDBC 4.5 API</h3>
```
The title is consistent with the other updates but I'm wondering why it hasn't been switched to "module java.sql".

AlanBateman commented on src/java.sql/share/classes/java/sql/Connection.java:
```diff
@@ -1676,5 +1676,268 @@ default void setShardingKey(ShardingKey shardingKey, ShardingKey superShardingKe
+     * @implNote
+     * JDBC driver implementations may need to provide their own implementation
+     * of this method in order to meet the requirements of the underlying
+     * datasource.
```
I realize this is how Statement.enquoteLiteral is specified since its introduction in JDK 9 but I can't see how a JDBC driver has any leeway to effect a different outcome. The spec is very clear that it quotes the input and replaces any single quotes with two single quotes. Do you have any examples where a "requirement" from a datasource could result in a different outcome? I could imagine a driver implementation doing its own implementation for performance reasons but not sure beyond that.

AlanBateman commented on src/java.sql/share/classes/java/sql/Connection.java:
```diff
@@ -1676,5 +1676,268 @@ default void setShardingKey(ShardingKey shardingKey, ShardingKey superShardingKe
+     * <ul>
+     * <li>Return the original value if {@code alwaysDelimit} is
+     * {@code false}</li>
+     * <li>Return a delimited identifier if {@code alwaysDelimit} is
```
Is there somewhere that defines "delimited identifier" that this could link to? I think it means double quotes but at the same time it seems that this may not be supported by the JDBC driver or datasource. This makes me wonder if the default implementation should throw SQLFeatureNotSupportedException as I assume something is broken if this method is returning a quoted result that is not supported.
